oauth2_http/java/com/google/auth/oauth2/AwsRequestSigner.java

```diff
@@ -0,0 +1,381 @@
+      canonicalHeaders.append(headerName).append(":").append(headers.get(headerName)).append("\n");
+    }
+    canonicalRequest.append(canonicalHeaders).append("\n");
+
+    // Append the signed headers.
+    canonicalRequest.append(Joiner.on(';').join(sortedHeaderNames)).append("\n");
+
+    // Append the hashed request payload.
+    canonicalRequest.append(getHexEncodedSha256Hash(requestPayload.getBytes()));
+
+    // Return the hashed canonical request.
+    return getHexEncodedSha256Hash(canonicalRequest.toString().getBytes());
+  }
+
+  /** Task 2: Create a string to sign for Signature Version 4. */
+  private String createStringToSign(
+      String canonicalRequestHash, String xAmzDate, String credentialScope) {
+    return HASHING_ALGORITHM
+        + "\n"
+        + xAmzDate
+        + "\n"
+        + credentialScope
+        + "\n"
+        + canonicalRequestHash;
+  }
+
+  /**
+   * Task 3: Calculate the signature for AWS Signature Version 4.
+   *
+   * @param date the date used in the hashing process in YYYYMMDD format
+   */
+  private String calculateAwsV4Signature(
+      String serviceName, String secret, String date, String region, String stringToSign) {
+    byte[] kDate = sign(("AWS4" + secret).getBytes(UTF_8), date.getBytes(UTF_8));
+    byte[] kRegion = sign(kDate, region.getBytes(UTF_8));
+    byte[] kService = sign(kRegion, serviceName.getBytes(UTF_8));
+    byte[] kSigning = sign(kService, AWS_REQUEST_TYPE.getBytes(UTF_8));
+    return BaseEncoding.base16().lowerCase().encode(sign(kSigning, stringToSign.getBytes(UTF_8)));
+  }
+
+  /** Task 4: Format the signature to be added to the HTTP request. */
+  private String generateAuthorizationHeader(
+      List<String> sortedHeaderNames,
+      String accessKeyId,
+      String credentialScope,
+      String signature) {
+    return String.format(
+        "%s Credential=%s/%s, SignedHeaders=%s, Signature=%s",
+        HASHING_ALGORITHM,
+        accessKeyId,
+        credentialScope,
+        Joiner.on(';').join(sortedHeaderNames),
+        signature);
+  }
+
+  private Map<String, String> getCanonicalHeaders(String date) {
+    Map<String, String> headers = new HashMap<>();
+    headers.put("host", uri.getHost());
+
+    // Only add the date if it hasn't been specified through the "date" header.
+    if (!additionalHeaders.containsKey("date")) {
+      headers.put("x-amz-date", date);
+    }
+
+    if (awsSecurityCredentials.getToken() != null && !awsSecurityCredentials.getToken().isEmpty()) {
+      headers.put("x-amz-security-token", awsSecurityCredentials.getToken());
+    }
+
+    // Add all additional headers.
+    for (String key : additionalHeaders.keySet()) {
+      // Header keys need to be lowercase.
+      headers.put(key.toLowerCase(Locale.US), additionalHeaders.get(key));
+    }
+    return headers;
+  }
+
+  private AwsDates getDates() {
+    if (additionalHeaders.containsKey("date")) {
+      return AwsDates.fromDateHeader(additionalHeaders.get("date"));
```

Comment:
Assuming the headers are read from a HTTP request to an external server, then a checked exception is expected here since the format is not guaranteed.